File: tiff_reader/tiff_reader.c
#include <assert.h>
#include <fcntl.h>
#include <float.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <unistd.h>

typedef uint8_t u8;
typedef uint16_t u16;
typedef int16_t i16;
typedef uint32_t u32;
typedef uint64_t u64;
typedef float f32;
typedef double f64;
typedef u32 b32;

#define READ_U16(map, offset) (*(u16 *)((map) + (offset)))
#define READ_I16(map, offset) (*(i16 *)((map) + (offset)))
#define READ_U32(map, offset) (*(u32 *)((map) + (offset)))
#define READ_F32(map, offset) (*(f32 *)((map) + (offset)))
#define READ_F64(map, offset) (*(f64 *)((map) + (offset)))

#define MIN(a, b) (((a) < (b)) ? (a) : (b))
#define MAX(a, b) (((a) > (b)) ? (a) : (b))
#define ABS(a) (((a) < 0) ? -(a) : (a))

enum TiffType {
  TIFF_SHORT = 3,
  TIFF_LONG = 4,
  TIFF_FLOAT = 11,
  TIFF_DOUBLE = 12,
};

enum SampleType {
  SAMPLE_UNSIGNED_INT = 1,
  SAMPLE_SIGNED_INT = 2,
  SAMPLE_FLOAT = 3,
  SAMPLE_UNDEFINED = 4,
};

typedef struct {
  u16 tag, type;
  u32 count, value_offset;
} ifd_entry;

ifd_entry read_entry(u8 *map, u32 *offset);

typedef struct {
  u32 capacity, length;
  enum TiffType dtype;
  u16 bytesize;
  u8 *data;
} vector;

u16 get_byte_size(enum TiffType dtype);
void vector_free(vector *vec);
vector *vector_from_slice(u8 *map, ifd_entry entry);
u32 vector_get_u32(vector *vec, u32 i);
f64 vector_get_f64(vector *vec, u32 i);

typedef struct {
  u32 image_width, image_length;
  u16 bits_per_sample;
  u16 compression;
  u16 photometric_interpretation;
  u16 samples_per_pixel;
  vector *strip_offsets;
  u32 rows_per_strip;
  u16 planar_configuration;
  u16 sample_format;
  vector *strip_byte_counts;
  u8 *projection;
  vector *model_tie_points;
  vector *model_pixel_scale_tag;
  vector *model_transformation_tag;
  vector *geo_double_params_tag;
} tiff_ifd;

tiff_ifd *ifd_init();
void parse_ifd_entry(tiff_ifd *ifd, u8 *map, u32 *offset);
void free_ifd(tiff_ifd *ifd);

typedef struct {
  u8 *map;
  tiff_ifd *ifd;
  f64 *x, *y;
  void *data;
} tiff_dataset;

tiff_dataset *read_tiff(u8 *map);
void tiff_load_data(tiff_dataset *tif);
void free_tiff(tiff_dataset *tif);
u32 searchsorted(f64 *values, f64 key, u32 n);

int main(int argc, char *argv[]) {
  if (argc < 2) {
    fprintf(stderr, "Input file must be provided\n");
    return EXIT_FAILURE;
  }
  int fd = open(argv[1], O_RDONLY);
  if (fd == -1) {
    fprintf(stderr, "Could not read file\n");
    return EXIT_FAILURE;
  }
  struct stat stat_buf;
  if (fstat(fd, &stat_buf) == -1) {
    fprintf(stderr, "Could not get file stats\n");
    return EXIT_FAILURE;
  }
  u8 *map = (u8 *)mmap(NULL, stat_buf.st_size, PROT_READ, MAP_PRIVATE, fd, 0);

  tiff_dataset *tif = read_tiff(map);
  if (tif == NULL)
    return EXIT_FAILURE;
  printf("%s\n", tif->ifd->projection);
  tiff_load_data(tif);
  u32 size = tif->ifd->image_length * tif->ifd->image_width;
  switch (tif->ifd->sample_format) {
  case SAMPLE_UNSIGNED_INT:
    printf("data[%u]=%hu\n", size - 1, *((u16 *)tif->data + size - 1));
    break;
  case SAMPLE_SIGNED_INT:
    printf("data[%u]=%d\n", size - 1, *((i16 *)tif->data + size - 1));
    break;
  case SAMPLE_FLOAT:
    printf("data[%u]=%f\n", size - 1, *((f32 *)tif->data + size - 1));
    break;
  }
  free_tiff(tif);

  munmap(map, stat_buf.st_size);

  return EXIT_SUCCESS;
}

void parse_int(u32 *dst, u8 *map, u16 type, u32 *offset) {
  switch (type) {
  case TIFF_SHORT:
    *dst = (u32)READ_U16(map, *offset);
    *offset += 2;
    break;
  case TIFF_LONG:
    *dst = READ_U32(map, *offset);
    *offset += 4;
    break;
  default:
    assert(0);
    break;
  }
}

u16 get_byte_size(enum TiffType dtype) {
  switch (dtype) {
  case TIFF_SHORT:
    return sizeof(u16);
  case TIFF_LONG:
    return sizeof(u32);
  case TIFF_FLOAT:
    return sizeof(f32);
  case TIFF_DOUBLE:
    return sizeof(f64);
  }
}

vector *vector_from_slice(u8 *map, ifd_entry entry) {
  vector *vec = malloc(sizeof(vector));
  u16 bytesize = get_byte_size(entry.type);
  vec->capacity = entry.count;
  vec->length = entry.count;
  vec->bytesize = bytesize, vec->dtype = entry.type,
  vec->data = (u8 *)malloc((u32)bytesize * entry.count);
  memcpy(vec->data, map + entry.value_offset, bytesize * entry.count);
  return vec;
}

void vector_free(vector *vec) {
  if (vec == NULL)
    return;
  free(vec->data);
  free(vec);
}

u32 vector_get_u32(vector *vec, u32 i) {
  return READ_U32(vec->data, i * vec->bytesize);
}

f64 vector_get_f64(vector *vec, u32 i) {
  return READ_F64(vec->data, i * vec->bytesize);
}

ifd_entry read_entry(u8 *map, u32 *offset) {
  ifd_entry entry = {.tag = READ_U16(map, *offset),
                     .type = READ_U16(map, *offset + 2),
                     .count = READ_U32(map, *offset + 4),
                     .value_offset = READ_U32(map, *offset + 8)};
  *offset += 12;
  return entry;
}

void parse_ifd_entry(tiff_ifd *ifd, u8 *map, u32 *offset) {
  ifd_entry entry = read_entry(map, offset);
  switch (entry.tag) {
  case 256:
    ifd->image_width = entry.value_offset;
    break;
  case 257:
    ifd->image_length = entry.value_offset;
    break;
  case 258:
    ifd->bits_per_sample = (u16)entry.value_offset;
    break;
  case 259:
    ifd->compression = (u16)entry.value_offset;
    break;
  case 262:
    ifd->photometric_interpretation = (u16)entry.value_offset;
    break;
  case 273:
    ifd->strip_offsets = vector_from_slice(map, entry);
    break;
  case 277:
    ifd->samples_per_pixel = (u16)entry.value_offset;
    break;
  case 278:
    ifd->rows_per_strip = entry.value_offset;
    break;
  case 279:
    ifd->strip_byte_counts = vector_from_slice(map, entry);
    break;
  case 284:
    ifd->planar_configuration = (u16)entry.value_offset;
    break;
  case 339:
    ifd->sample_format = (u16)entry.value_offset;
    break;
  case 33922:
    ifd->model_tie_points = vector_from_slice(map, entry);
    break;
  case 33550:
    ifd->model_pixel_scale_tag = vector_from_slice(map, entry);
    break;
  case 34264:
    assert(entry.count == 16);
    ifd->model_transformation_tag = vector_from_slice(map, entry);
    break;
  case 34735:
    // GeoKeys -> do nothing with it so far...
    // u16 n_key = READ_U16(map, entry.value_offset + 6);
    // u32 start = entry.value_offset + 8;
    // for (u16 i = 0; i < n_key; ++i) {
    //   printf("id=%hu, tag=%hu, count=%hu, value/offset=%hu\n",
    //          READ_U16(map, start), READ_U16(map, start + 2),
    //          READ_U16(map, start + 4), READ_U16(map, start + 6));
    //   start += 8;
    // }
    break;
  case 34736:
    ifd->geo_double_params_tag = vector_from_slice(map, entry);
    break;
  case 34737:
    ifd->projection = (u8 *)malloc(entry.count);
    memcpy(ifd->projection, map + entry.value_offset, entry.count);
    break;
  default:
    printf("Unknown IFD entry: Tag=%hu, Type=%hu, Count=%u, Offset/Value=%u\n",
           entry.tag, entry.type, entry.count, entry.value_offset);
    break;
  }
}

void free_ifd(tiff_ifd *ifd) {
  vector_free(ifd->strip_byte_counts);
  vector_free(ifd->strip_offsets);
  vector_free(ifd->model_tie_points);
  vector_free(ifd->model_pixel_scale_tag);
  vector_free(ifd->model_transformation_tag);
  vector_free(ifd->geo_double_params_tag);
  free(ifd->projection);
  free(ifd);
}

tiff_ifd *ifd_init() {
  tiff_ifd *ifd = (tiff_ifd *)malloc(sizeof(tiff_ifd));
  ifd->model_tie_points = NULL;
  ifd->model_pixel_scale_tag = NULL;
  ifd->model_transformation_tag = NULL;
  ifd->geo_double_params_tag = NULL;
  return ifd;
}

tiff_dataset *read_tiff(u8 *map) {
  u16 endianness = READ_U16(map, 0);
  if (endianness != *(u16 *)"II") {
    fprintf(stderr, "Current implementatio for little endian only\n");
    return NULL;
  }
  u16 magic_number = READ_U16(map, 2);
  if (magic_number == 43) {
    fprintf(stderr, "Current implementation does not support BigTIFF format\n");
    return NULL;
  }
  assert(READ_U16(map, 2) == 42);
  u32 offset = READ_U32(map, 4);
  u16 n_entry = READ_U16(map, offset);

  tiff_ifd *ifd = ifd_init();
  offset += 2;
  for (u16 i = 0; i < n_entry; ++i) {
    parse_ifd_entry(ifd, map, &offset);
  }
  offset = READ_U32(map, offset);
  assert(offset == 0);
  assert(ifd->sample_format != SAMPLE_UNDEFINED);

  tiff_dataset *tif = malloc(sizeof(tiff_dataset));
  tif->map = map;
  tif->ifd = ifd;
  tif->data = NULL;
  tif->x = malloc(sizeof(f64) * ifd->image_width);
  tif->y = malloc(sizeof(f64) * ifd->image_length);

  if (ifd->model_pixel_scale_tag && ifd->model_tie_points) {
    printf("%d\n", ifd->model_pixel_scale_tag->length);
    assert(ifd->model_pixel_scale_tag->length == 3);
    assert(ifd->model_tie_points->length == 6);
    // Assume tie point is the upper left corner
    f64 *tie_points = (f64 *)ifd->model_tie_points->data;
    f64 *scale_tag = (f64 *)ifd->model_pixel_scale_tag->data;
    f64 dx = scale_tag[0];
    tif->x[0] = tie_points[3];
    for (u32 ix = 1; ix < ifd->image_width; ++ix) {
      tif->x[ix] = tif->x[ix - 1] + dx;
    }
    f64 dy = scale_tag[1];
    tif->y[0] = tie_points[4];
    for (u32 iy = 1; iy < ifd->image_length; ++iy) {
      tif->y[iy] = tif->y[iy - 1] - dy;
    }
    return tif;
  }
  if (ifd->model_transformation_tag) {
    f64 *trans = (f64 *)ifd->model_transformation_tag->data;
    assert(ABS(trans[1]) < DBL_EPSILON && ABS(trans[4]) < DBL_EPSILON);
    for (u64 i = 0; i < (u64)ifd->image_width; ++i) {
      tif->x[i] = trans[3] + trans[0] * (f64)i;
    }
    for (u64 i = 0; i < (u64)ifd->image_length; ++i) {
      tif->y[i] = trans[7] + trans[5] * (f64)i;
    }
    return tif;
  }

  fprintf(stderr, "Invalid transformation\n");
  free_tiff(tif);
  return NULL;
}

u32 searchsorted(f64 *values, f64 key, u32 n) {
  f64 step = values[1] - values[0];
  u32 index = (u32)((key - values[0]) / step);
  return MAX(0, MIN(index, n - 2));
}

void tiff_load_data(tiff_dataset *tif) {
  u32 nx = tif->ifd->image_width;
  u32 ny = tif->ifd->image_length;
  u32 size = (u32)tif->ifd->bits_per_sample / 8;
  tif->data = malloc(size * nx * ny);
  u32 n_stripe = tif->ifd->strip_offsets->length;
  u32 *strip_offsets = (u32 *)tif->ifd->strip_offsets->data;
  u16 *strip_byte_counts = (u16 *)tif->ifd->strip_byte_counts->data;
  u32 pixel = 0;
  for (u32 stripe = 0; stripe < n_stripe; ++stripe) {
    u32 offset = strip_offsets[stripe];
    u32 byte_count = (u32)strip_byte_counts[stripe];
    for (u32 i = 0; i < byte_count; i += size) {
      switch (tif->ifd->sample_format) {
      case SAMPLE_UNSIGNED_INT:
        *((u16 *)tif->data + pixel++) = READ_U16(tif->map, offset + i);
        break;
      case SAMPLE_SIGNED_INT:
        *((i16 *)tif->data + pixel++) = READ_I16(tif->map, offset + i);
        break;
      case SAMPLE_FLOAT:
        *((f32 *)tif->data + pixel++) = READ_F32(tif->map, offset + i);
        break;
      default:
        fprintf(stderr, "Unexpected sample format");
        free(tif->data);
        return;
      }
      // printf("pixel=%u/%u\n", pixel, nx * ny);
    }
  }
}

void free_tiff(tiff_dataset *tif) {
  free_ifd(tif->ifd);
  if (tif->data)
    free(tif->data);
  free(tif->x);
  free(tif->y);
  free(tif);
}

File: tiff_reader/tiff_reader.zig
const std = @import("std");
const builtin = @import("builtin");

const Endian = std.builtin.Endian;
const native_endian = builtin.cpu.arch.endian();

const TiffDataType = enum(u16) {
    short = 3,
    long = 4,
    float = 11,
    double = 12,
};

const TiffError = error{ InvalidFirstBytes, BadMagicNumber, InvalidDataType, UnsupportedBigTiff, UnknownTag, TooManyIFDs, InvalidTransformation };

const IfdEntry = struct {
    tag: u16,
    field_type: u16,
    count: u32,
    value_offset: u32,
};

const TiffIfd = struct {
    image_width: u32,
    image_length: u32,
    bits_per_sample: u16,
    compression: u16,
    photometric_interpretation: u16,
    samples_per_pixel: u16,
    strip_offsets: []u32,
    rows_per_strip: u32,
    planar_configuration: u16,
    sample_format: u16,
    strip_byte_counts: []u32,
    projection: []u8,
    model_tie_points: ?[]f64,
    model_pixel_scale_tag: ?[]f64,
    model_transformation_tag: ?[16]f64,
    geo_double_params_tag: ?[]f64,

    pub fn init(allocator: std.mem.Allocator) !*TiffIfd {
        var ifd = try allocator.create(TiffIfd);
        ifd.model_tie_points = null;
        ifd.model_pixel_scale_tag = null;
        ifd.model_transformation_tag = null;
        ifd.geo_double_params_tag = null;
        return ifd;
    }
};

const TiffDataset = struct {
    ifd: *TiffIfd,
    x: []f64,
    y: []f64,
    data: ?[]f32,

    pub fn from_ifd(allocator: std.mem.Allocator, ifd: *TiffIfd) !TiffDataset {
        const nx = ifd.image_width;
        const ny = ifd.image_length;
        var x = try allocator.alloc(f64, nx);
        var y = try allocator.alloc(f64, ny);
        const eps = std.math.floatEps(f64);
        if (ifd.model_transformation_tag) |trans| {
            if (@abs(trans[1]) > eps or @abs(trans[4]) > eps) {
                return TiffError.InvalidTransformation;
            }
            for (0..nx) |i| {
                x[i] = trans[3] + trans[0] * @as(f64, @floatFromInt(i));
            }
            for (0..ny) |i| {
                y[i] = trans[7] + trans[5] * @as(f64, @floatFromInt(i));
            }
            return TiffDataset{ .ifd = ifd, .x = x, .y = y, .data = null };
        }
        if (ifd.model_tie_points != null and ifd.model_pixel_scale_tag != null) {
            // Assume upper left corner is provided
            const pixel_scale = ifd.model_pixel_scale_tag.?;
            const tie_points = ifd.model_tie_points.?;
            if (tie_points.len != 6) {
                std.debug.print("model_tie_points has unexpected size", .{});
                return TiffError.InvalidTransformation;
            }
            const t0: u64 = @intFromFloat(tie_points[0]);
            const t1: u64 = @intFromFloat(tie_points[1]);
            if (t0 != 0 or t1 != 0) {
                std.debug.print("model_tie_point is not (0, 0)", .{});
                return TiffError.InvalidTransformation;
            }
            x[0] = tie_points[3];
            for (1..nx) |i| {
                x[i] = x[i - 1] + pixel_scale[0];
            }
            y[0] = tie_points[4];
            for (1..ny) |i| {
                y[i] = y[i - 1] - pixel_scale[1];
            }
            return TiffDataset{ .ifd = ifd, .x = x, .y = y, .data = null };
        }
        return TiffError.InvalidTransformation;
    }

    pub fn get_extent(self: *const TiffDataset) [4]f64 {
        const nx = self.ifd.image_width;
        const ny = self.ifd.image_length;
        var xmin = self.x[0];
        var xmax = self.x[nx - 1];
        if (xmin > xmax) {
            std.mem.swap(f64, &xmin, &xmax);
        }
        var ymax = self.y[0];
        var ymin = self.y[ny - 1];
        if (ymin > ymax) {
            std.mem.swap(f64, &ymin, &ymax);
        }
        return .{ xmin, xmax, ymin, ymax };
    }

    pub fn load_data(self: *TiffDataset, reader: *TiffReader) !void {
        var data = try reader.allocator.alloc(f32, self.ifd.image_length * self.ifd.image_width);
        var pixel: usize = 0;
        for (self.ifd.strip_offsets, 0..) |offset, strip| {
            reader.offset = @intCast(offset);
            const nmax = self.ifd.strip_byte_counts[strip] / 4;
            for (0..nmax) |_| {
                data[pixel] = reader.read_scalar(f32);
                pixel += 1;
            }
        }
        self.data = data;
    }
};

const TiffReader = struct {
    allocator: std.mem.Allocator,
    data: []const u8,
    offset: usize,
    endian: Endian,

    pub fn new(allocator: std.mem.Allocator, data: []const u8) !TiffReader {
        if (std.mem.eql(u8, data[0..2], "II")) {
            return TiffReader{ .allocator = allocator, .data = data, .offset = 2, .endian = .little };
        }
        if (std.mem.eql(u8, data[0..2], "MM")) {
            return TiffReader{ .allocator = allocator, .data = data, .offset = 2, .endian = .big };
        }
        return TiffError.InvalidFirstBytes;
    }

    pub fn read_scalar(self: *TiffReader, comptime T: type) T {
        const shift = @sizeOf(T);
        const buffer: *const [@sizeOf(T)]u8 = @ptrCast(self.data[self.offset .. self.offset + shift]);
        self.offset += shift;
        if (@typeInfo(T) == .float) {
            const value: @Type(.{ .int = .{ .signedness = .unsigned, .bits = @typeInfo(T).float.bits } }) = @bitCast(buffer.*);
            return if (self.endian == native_endian) @bitCast(value) else @bitCast(@byteSwap(value));
        }
        const value: T = @bitCast(buffer.*);
        return if (self.endian == native_endian) value else @byteSwap(value);
    }

    pub fn read_int_vector(self: *TiffReader, comptime T: type, entry: *const IfdEntry) ![]T {
        if (@typeInfo(T) != .int) {
            return TiffError.InvalidDataType;
        }
        var vec = try self.allocator.alloc(T, entry.count);
        const current = self.offset;
        self.offset = @intCast(entry.value_offset);
        for (0..entry.count) |i| {
            switch (entry.field_type) {
                @intFromEnum(TiffDataType.short) => {
                    vec[i] = @intCast(self.read_scalar(u16));
                },
                @intFromEnum(TiffDataType.long) => {
                    vec[i] = @intCast(self.read_scalar(u32));
                },
                else => {
                    return TiffError.InvalidDataType;
                },
            }
        }
        self.offset = current;
        return vec;
    }

    pub fn read_float_vector(self: *TiffReader, comptime T: type, entry: *const IfdEntry) ![]T {
        if (@typeInfo(T) != .float) {
            return TiffError.InvalidDataType;
        }
        var vec = try self.allocator.alloc(T, entry.count);
        const current = self.offset;
        self.offset = @intCast(entry.value_offset);
        for (0..entry.count) |i| {
            switch (entry.field_type) {
                @intFromEnum(TiffDataType.float) => {
                    vec[i] = @floatCast(self.read_scalar(f32));
                },
                @intFromEnum(TiffDataType.double) => {
                    vec[i] = @floatCast(self.read_scalar(f64));
                },
                else => {
                    return TiffError.InvalidDataType;
                },
            }
        }
        self.offset = current;
        return vec;
    }

    pub fn read_ifd_entry(self: *TiffReader) IfdEntry {
        const tag = self.read_scalar(u16);
        const field_type = self.read_scalar(u16);
        const count = self.read_scalar(u32);
        const value_offset = self.read_scalar(u32);
        return IfdEntry{ .tag = tag, .field_type = field_type, .count = count, .value_offset = value_offset };
    }

    pub fn process_ifd_entry(self: *TiffReader, ifd: *TiffIfd) !void {
        const entry = self.read_ifd_entry();
        switch (entry.tag) {
            256 => {
                ifd.image_width = entry.value_offset;
            },
            257 => {
                ifd.image_length = entry.value_offset;
            },
            258 => {
                ifd.bits_per_sample = @intCast(entry.value_offset);
            },
            259 => {
                ifd.compression = @intCast(entry.value_offset);
            },
            262 => {
                ifd.photometric_interpretation = @intCast(entry.value_offset);
            },
            273 => {
                ifd.strip_offsets = try self.read_int_vector(u32, &entry);
            },
            277 => {
                ifd.samples_per_pixel = @intCast(entry.value_offset);
            },
            278 => {
                ifd.rows_per_strip = entry.value_offset;
            },
            279 => {
                ifd.strip_byte_counts = try self.read_int_vector(u32, &entry);
            },
            284 => {
                ifd.planar_configuration = @intCast(entry.value_offset);
            },
            339 => {
                ifd.sample_format = @intCast(entry.value_offset);
            },
            33922 => {
                ifd.model_tie_points = try self.read_float_vector(f64, &entry);
            },
            33550 => {
                ifd.model_pixel_scale_tag = try self.read_float_vector(f64, &entry);
            },
            34264 => {
                const vec = try self.read_float_vector(f64, &entry);
                var trans: [16]f64 = undefined;
                @memcpy(&trans, vec);
                ifd.model_transformation_tag = trans;
                self.allocator.free(vec);
            },
            34735 => {},
            34736 => {
                ifd.geo_double_params_tag = try self.read_float_vector(f64, &entry);
            },
            34737 => {
                const start = entry.value_offset;
                const stop = start + entry.count;
                ifd.projection = try self.allocator.dupe(u8, self.data[start..stop]);
            },
            else => {
                return TiffError.UnknownTag;
            },
        }
    }

    pub fn read_tiff(self: *TiffReader) !TiffDataset {
        self.offset = 2;
        const magic = self.read_scalar(u16);
        if (magic == 43) {
            return TiffError.UnsupportedBigTiff;
        } else if (magic != 42) {
            std.debug.print("Big Tiff format currently not supported\n", .{});
            return TiffError.BadMagicNumber;
        }

        self.offset = @intCast(self.read_scalar(u32));
        const n_entry = self.read_scalar(u16);
        const ifd = try TiffIfd.init(self.allocator);
        for (0..n_entry) |_| {
            try self.process_ifd_entry(ifd);
        }
        if (self.read_scalar(u32) != 0) {
            std.debug.print("Found more than 1 IFD\n", .{});
            return TiffError.TooManyIFDs;
        }
        if (ifd.sample_format != 3) {
            std.debug.print("Only deals with float sample at the moment\n", .{});
            return TiffError.InvalidDataType;
        }
        return try TiffDataset.from_ifd(self.allocator, ifd);
    }
};

pub fn main() !void {
    if (std.os.argv.len != 2) {
        return error.InvalidArgument;
    }
    const cwd = std.fs.cwd();
    const file_name: [:0]const u8 = std.mem.span(std.os.argv[1]);
    const file = try cwd.openFile(file_name, .{});
    const stat = try std.posix.fstat(file.handle);
    var arena: std.heap.ArenaAllocator = .init(std.heap.c_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    const map: []u8 = try std.posix.mmap(null, @intCast(stat.size), std.posix.PROT.READ, .{ .TYPE = .SHARED }, file.handle, 0);
    var tiff_reader: TiffReader = try TiffReader.new(allocator, map);
    var tiff_data = try tiff_reader.read_tiff();
    std.debug.print("{any}\n", .{tiff_data.get_extent()});
    try tiff_data.load_data(&tiff_reader);
    const size = tiff_data.data.?.len;
    std.debug.print("data[{}]={}\n", .{ size - 1, tiff_data.data.?[size - 1] });
}
